### PR TITLE
[WIP] Belt Hell

### PR DESCRIPTION
**Pull request by francinum**
:cl: Francinum
add: Belt Hell sorting machinery, Not actually mapped in.
/:cl:
### Adds Goon's belt-based sorting machinery.

Cargo Router: Sorting machines, and the backbone of the network. 
https://file.house/jmV2.webm
Mass Driver Loader: Send things from one place to another, possibly through space. https://file.house/KelT.webm
Barcode Computer: A big, stationary variant of the destination tagger that's made to support routers.

Properly compiles, Code likely needs to be reworked.

### CODE REVIEW

**Comment by KorPhaeron** [comment]
It's a "someone opens a PR without explaining what the hell it is" episode

**Comment by Shadowlight213** [comment]
@KorPhaeron Belt hell is the name of Goon's conveyor mail system.

**Comment by MMMiracles** [comment]
I'm almost positive its the belt delivery system that goon's maps use, with a giant wrap-around belt system instead of just using disposals for trash/transporting shit.

**Comment by francinum** [comment]
Correct.

File: code/game/machinery/computer/barcode.dm [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		dat += "<b><A href='?src=\ref[src];print=[I]'>[I]</A></b><BR>"
+	user.machine = src
+	user << browse("<TITLE>Barcode Computer</TITLE><BR>[dat]", "window=bc_computer;size=400x300")
+	onclose(user, "bc_computer")
+	return
+
+
+/obj/machinery/computer/barcode/attackby(var/obj/item/I as obj, user as mob) //This is mostly goonconomy. I'll get to it later.
+	attack_hand(user)
+	return
```
Comment (phil235) [inline code/game/machinery/computer/barcode.dm]: you can't attack the thing with a weapon?

**Comment by KorPhaeron** [comment]
I know it's something cargo related from goon, but that doesn't help explain much if you're not familiar with it

File: code/game/machinery/computer/barcode.dm (hunk 1 above)
Comment (francinum) [inline code/game/machinery/computer/barcode.dm]: I'll swap it out.

**Comment by francinum** [comment]
@KorPhaeron I'm getting some footage now.

**Comment by kevinz000** [comment]
This might be interesting, but would take a shitload of work to implement into existing maps.

**Comment by Shadowlight213** [comment]
@kevinz000 existing maps? Yeah. downright impossible. Would require a complete and utter redesign of maint and departmental placement in general. Disposals goes under floors, which means it can just be placed in the hallways or along atmos and power routes. Belt hell is actually on top of the floor, so it needs its own paths mapped out.

New maps? Less impossible.
Belt hell is something needs to be considered and integrated in the first stages of the map design. Maint has to be expanded enough, and departments need to be placed such that the conveyors all fit. However, when it's actually added, belt hell is a lot of fun just to ride around in.

I know @francinum has been mentioning that they are working on some sort of goon inspired map for a while now, so it's likely that they have designed belt hell into said map already.

**Comment by kevinz000** [comment]
The last time I rode belt hell I ended up int he crusher
Please don't make it instagib :^)

**Comment by francinum** [comment]
@kevinz000 I make no promises.

**Comment by Iamgoofball** [comment]
BELT HELL NOW

BELT HELL FOREVER

**Comment by phil235** [comment]
cargo_router and launcher_loader need a Destroy() to clear references. Since this is a goon port there's probably a lot of things to change to not cause issue on tg.

**Comment by optimumtact** [comment]
no goon features and this is just bloat.

**Comment by MMMiracles** [comment]
tangerine go home

**Comment by optimumtact** [comment]
ez bait.

File: code/game/machinery/belts/launcherloaders.dm [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+	var/id = "null"
+	var/operating = 0
+	var/driver_operating = 0
+	var/trash = 0
+	var/obj/machinery/door/poddoor/door = null
+
+/obj/machinery/launcher_loader/New()
+	..()
+	spawn(5)
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: why is this spawn here?

File: code/game/machinery/belts/launcherloaders.dm [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+				for(var/obj/machinery/mass_driver/D2 in drivers)
+					if(D2.id == src.id)
+						driver = D2
+						break
+				if(!driver) driver = pick(drivers)
+			else
+				driver = pick(drivers)
+			src.dir = get_dir(src,driver)
+
+/obj/machinery/launcher_loader/proc/destroy()
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: Destroy, not destroy

File: code/game/machinery/belts/launcherloaders.dm [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+	if (door)
+		door = null
+	return ..()
+
+/obj/machinery/launcher_loader/proc/activate()
+	if(operating || !isturf(src.loc)) return
+	operating = 1
+	flick("launcher_loader_1",src)
+	playsound(src, "goon/sound/effects/pump.ogg",50, 1)
+	sleep(3)
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: you need to check you still exist and are not qdeleted after this sleep

File: code/game/machinery/belts/launcherloaders.dm [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+					if(D.id == driver.id)
+						D.drive()
+/obj/machinery/launcher_loader/process()
+	if(!operating && !driver_operating)
+		var/drive = 0
+		for(var/atom/movable/M in src.loc)
+			if(M == src || M.anchored) continue
+			drive = 1
+			break
+		if(drive) activate()
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: codebase standard is methods don't go on the same line as ifs

Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: same goes for a bunch of locations

File: code/game/machinery/belts/launcherloaders.dm [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+		step(AM,src.dir)
+	operating = 0
+	handle_driver()
+
+/obj/machinery/launcher_loader/proc/handle_driver()
+	if(driver && !driver_operating)
+		driver_operating = 1
+
+	spawn(0)
+		for(var/obj/machinery/door/poddoor/P)
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: is this  a loop over contents? This doesn't even have a list associated, how the fuck does this code even work

Comment (francinum) [inline code/game/machinery/belts/launcherloaders.dm]: prayers and magic

File: code/game/machinery/belts/launcherloaders.dm [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+			if (P.id == driver.id)
+				door = P
+				if (door)
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: Why would you check that the door exists at this point?

File: code/game/machinery/belts/launcherloaders.dm [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+		driver = null
+	return ..()
+
+/obj/machinery/cargo_router/proc/activate()
+	if(operating || !isturf(src.loc)) return
+	operating = 1
+
+	var/next_dest = null
+
+	for(var/atom/movable/AM in src.loc)
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: get the turf instead with var/turf/target = get_turf(src)

also removes the need for the isturf check in the if further up (just check if(target))

File: code/game/machinery/belts/launcherloaders.dm [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+		if(AM.anchored || AM == src) continue
+		if(AM.delivery_destination && !next_dest)
+			if(destinations.Find(AM.delivery_destination))
+				next_dest = destinations[AM.delivery_destination]
+				break
+
+	if(next_dest) src.dir = next_dest
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: uncessary src's again

File: code/game/machinery/belts/launcherloaders.dm [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+	else src.dir = default_direction
+
+	flick("amdl_1",src)
+	playsound(src, "goon/sound/effects/pump.ogg",50, 1)
+	sleep(3)
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: you need to check src and !qdeleted after sleeps

File: code/game/machinery/belts/launcherloaders.dm [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+	operating = 0
+	handle_driver()
+
+/obj/machinery/cargo_router/proc/handle_driver()
+	if(driver && !driver_operating)
+		driver_operating = 1
+
+		spawn(0)
+			spawn(20)
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: more spawns that should use world.timers to reset vars

File: code/game/machinery/computer/barcode.dm [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+	if (..(user))
+		return
+
+	var/dat = ""
+	dat += "<b>Available Destinations:</b><BR>"
+	for(var/I in destinations)
+		dat += "<b><A href='?src=\ref[src];print=[I]'>[I]</A></b><BR>"
+	user.machine = src
+	user << browse("<TITLE>Barcode Computer</TITLE><BR>[dat]", "window=bc_computer;size=400x300")
+	onclose(user, "bc_computer")
```
Comment (optimumtact) [inline code/game/machinery/computer/barcode.dm]: I don't know if we support onclose

File: code/game/machinery/computer/barcode.dm [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+	return
+
+/obj/machinery/computer/barcode/Topic(href, href_list)
+	if (..(href, href_list))
+		return
+
+	if (href_list["print"] && !printing)
+		printing = 1
+		playsound(src.loc, "goon/sound/machines/printer_thermal.ogg", 50, 0)
+		sleep(28)
```
Comment (optimumtact) [inline code/game/machinery/computer/barcode.dm]: check after sleep you still exist and are not qdeleted.

A better way to do this one is just add_timer(src, 28, "print_sticker") and implement a print sticker method

File: code/game/machinery/computer/barcode.dm [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+	icon_state = "barcodesticker"
+	item_state = "paper"
+	var/destination = "QM Dock"
+
+/obj/item/barcodesticker/proc/attachTo(atom/movable/target)
+	if(get_dist(get_turf(target), get_turf(src)) <= 1 )
+		if(target==loc && target != usr)
+			return //Backpack or something
+		target.delivery_destination = destination
+		usr.visible_message("<span style=\"color:blue\">[usr] puts a [src.name] on [target].</span>")
```
Comment (optimumtact) [inline code/game/machinery/computer/barcode.dm]: pass user through explicitly, it's available in attack, afterattack and you can pass in usr through MouseDrop (we prefer to keep usr) confined to top level procs where possible.

File: code/game/machinery/computer/barcode.dm [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+/obj/item/barcodesticker
+	name = "Barcode Sticker"
+	desc = "A barcode sticker used in the cargo routing system."
+	icon = 'goon/icons/obj/barcode.dmi'
+	icon_state = "barcodesticker"
+	item_state = "paper"
+	var/destination = "QM Dock"
+
+/obj/item/barcodesticker/proc/attachTo(atom/movable/target)
+	if(get_dist(get_turf(target), get_turf(src)) <= 1 )
```
Comment (optimumtact) [inline code/game/machinery/computer/barcode.dm]: if(adjacent(target))

File: code/game/machinery/belts/launcherloaders.dm [hunk 16]
```diff
@@ -0,0 +1,186 @@
+/obj/machinery/launcher_loader
+	icon = 'goon/icons/obj/stationobjs.dmi'
+	icon_state = "launcher_loader_0"
+	name = "Automatic mass-driver loader (AMDL)"
+	desc = "An automated, hydraulic mass-driver loader."
+	density = 0
+	opacity = 0
+	layer = 2.6
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: this should use a layer define

File: code/game/machinery/belts/launcherloaders.dm [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+		for(var/obj/machinery/door/poddoor/P)
+			if (P.id == driver.id)
+				door = P
+				if (door)
+					door.open()
+			spawn(99)
+				if (door)
+					door.close() //this may need some adjusting still
+
+		spawn(door ? 55 : 20) driver_operating = 0
```
Comment (optimumtact) [inline code/game/machinery/belts/launcherloaders.dm]: it seems a lot of this can just be world.time timer vars in the process.

**Review by optimumtact — CHANGES_REQUESTED** [review]
needs work.

**Comment by optimumtact** [comment]
I hope this is old goon code

**Comment by francinum** [comment]
This is a special kind of shitcode.

File: code/game/machinery/belts/launcherloaders.dm (hunk 6 above)
Comment (AnturK) [inline code/game/machinery/belts/launcherloaders.dm]: For loops default to in world. Aka Bad.

File: code/game/machinery/belts/launcherloaders.dm (hunk 17 above)
Comment (Iamgoofball) [inline code/game/machinery/belts/launcherloaders.dm]: What the fuck is this

**Comment by KorPhaeron** [comment]
lmao gooncode

**Pull request closed without merging**